File: shell/src/main/java/com/sequenceiq/cloudbreak/shell/commands/ClusterCommands.java
package com.sequenceiq.cloudbreak.shell.commands;

import static com.sequenceiq.cloudbreak.shell.support.TableRenderer.renderSingleMap;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import javax.inject.Inject;

import org.springframework.shell.core.CommandMarker;
import org.springframework.shell.core.annotation.CliAvailabilityIndicator;
import org.springframework.shell.core.annotation.CliCommand;
import org.springframework.shell.core.annotation.CliOption;
import org.springframework.stereotype.Component;

import com.sequenceiq.cloudbreak.api.model.AmbariStackDetailsJson;
import com.sequenceiq.cloudbreak.api.model.ClusterRequest;
import com.sequenceiq.cloudbreak.api.model.ClusterResponse;
import com.sequenceiq.cloudbreak.api.model.ConfigStrategy;
import com.sequenceiq.cloudbreak.api.model.ConstraintJson;
import com.sequenceiq.cloudbreak.api.model.FileSystemRequest;
import com.sequenceiq.cloudbreak.api.model.FileSystemType;
import com.sequenceiq.cloudbreak.api.model.HostGroupAdjustmentJson;
import com.sequenceiq.cloudbreak.api.model.HostGroupJson;
import com.sequenceiq.cloudbreak.api.model.Status;
import com.sequenceiq.cloudbreak.api.model.StatusRequest;
import com.sequenceiq.cloudbreak.api.model.UpdateClusterJson;
import com.sequenceiq.cloudbreak.api.model.UpdateStackJson;
import com.sequenceiq.cloudbreak.client.CloudbreakClient;
import com.sequenceiq.cloudbreak.shell.completion.HostGroup;
import com.sequenceiq.cloudbreak.shell.model.CloudbreakContext;
import com.sequenceiq.cloudbreak.shell.model.Hints;
import com.sequenceiq.cloudbreak.shell.model.HostgroupEntry;
import com.sequenceiq.cloudbreak.shell.model.MarathonContext;
import com.sequenceiq.cloudbreak.shell.transformer.ExceptionTransformer;
import com.sequenceiq.cloudbreak.shell.transformer.ResponseTransformer;
import com.sequenceiq.cloudbreak.shell.util.CloudbreakShellUtil;

@Component
public class ClusterCommands implements CommandMarker {

    @Inject
    private CloudbreakContext context;
    @Inject
    private MarathonContext marathonContext;
    @Inject
    private CloudbreakClient cloudbreakClient;
    @Inject
    private ResponseTransformer responseTransformer;
    @Inject
    private CloudbreakShellUtil cloudbreakShellUtil;
    @Inject
    private ExceptionTransformer exceptionTransformer;

    @CliAvailabilityIndicator(value = "cluster create")
    public boolean isClusterCreateCommandAvailable() {
        return context.isBlueprintAvailable()
                && ((context.isStackAvailable() && context.getActiveHostGroups().size() == context.getHostGroups().keySet().size())
                        || (context.isMarathonMode() && marathonContext.isSelectedMarathonStackAvailable()
                && context.getActiveHostGroups().size() == marathonContext.getHostgroups().size()));
    }

    @CliAvailabilityIndicator(value = { "cluster show", "cluster stop", "cluster start" })
    public boolean isClusterShowCommandAvailable() {
        return context.isStackAvailable() || (context.isMarathonMode() && marathonContext.isSelectedMarathonStackAvailable());
    }

    @CliAvailabilityIndicator({ "cluster node --ADD", "cluster node --REMOVE" })
    public boolean isClusterNodeCommandAvailable() {
        return context.isStackAvailable() || (context.isMarathonMode() && marathonContext.isSelectedMarathonStackAvailable());
    }

    @CliAvailabilityIndicator({ "cluster fileSystem --DASH", "cluster fileSystem --GCS", "cluster fileSystem --WASB" })
    public boolean isClusterFileSystemCommandAvailable() {
        return context.isStackAvailable() && !context.isMarathonMode();
    }

    @CliCommand(value = "cluster fileSystem --DASH", help = "Set Windows Azure Blob Storage filesystem with DASH on cluster")
    public String setAzureRmFileSystem(
            @CliOption(key = "defaultFileSystem", mandatory = true, help = "Use as default filesystem") Boolean defaultFileSystem,
            @CliOption(key = "accountName", mandatory = true, help = "accountName of the DASH service") String accountName,
            @CliOption(key = "accountKey", mandatory = true, help = "access key of the DASH service") String accountKey) {
        context.setDefaultFileSystem(defaultFileSystem);
        context.setFileSystemType(FileSystemType.DASH);
        Map<String, Object> props = new HashMap<>();
        props.put("accountName", accountName);
        props.put("accountKey", accountKey);
        context.setFileSystemParameters(props);
        return "Windows Azure Blob Storage with DASH configured as the filesystem";
    }

    @CliCommand(value = "cluster fileSystem --WASB", help = "Set Windows Azure Blob Storage filesystem on cluster")
    public String setWasbFileSystem(
            @CliOption(key = "defaultFileSystem", mandatory = true, help = "Use as default filesystem") Boolean defaultFileSystem,
            @CliOption(key = "accountName", mandatory = true, help = "name of the storage account") String accountName,
            @CliOption(key = "accountKey", mandatory = true, help = "primary access key to the storage account") String accountKey) {
        context.setDefaultFileSystem(defaultFileSystem);
        context.setFileSystemType(FileSystemType.WASB);
        Map<String, Object> props = new HashMap<>();
        props.put("accountName", accountName);
        props.put("accountKey", accountKey);
        context.setFileSystemParameters(props);
        return "Windows Azure Blob Storage filesystem configured";
    }


    @CliCommand(value = "cluster fileSystem --GCS", help = "Set GCS fileSystem on cluster")
    public String setGcsFileSystem(
            @CliOption(key = "defaultFileSystem", mandatory = true, help = "Use as default fileSystem") Boolean defaultFileSystem,
            @CliOption(key = "projectId", mandatory = true, help = "projectId of the GCS") String projectId,
            @CliOption(key = "serviceAccountEmail", mandatory = true, help = "serviceAccountEmail of the GCS") String serviceAccountEmail,
            @CliOption(key = "privateKeyEncoded", mandatory = true, help = "privateKeyEncoded of the GCS") String privateKeyEncoded,
            @CliOption(key = "defaultBucketName", mandatory = true, help = "defaultBucketName of the GCS") String defaultBucketName) {
        context.setDefaultFileSystem(defaultFileSystem);
        context.setFileSystemType(FileSystemType.GCS);
        Map<String, Object> props = new HashMap<>();
        props.put("projectId", projectId);
        props.put("serviceAccountEmail", serviceAccountEmail);
        props.put("privateKeyEncoded", privateKeyEncoded);
        props.put("defaultBucketName", defaultBucketName);
        context.setFileSystemParameters(props);
        return "GCS filesystem configured";
    }

    @CliCommand(value = "cluster show", help = "Shows the cluster by stack id")
    public String configCluster() {
        try {
            String stackId = context.isMarathonMode() ? marathonContext.getSelectedMarathonStackId().toString() : context.getStackId();
            ClusterResponse clusterResponse = cloudbreakClient.clusterEndpoint().get(Long.valueOf(stackId));
            return renderSingleMap(responseTransformer.transformObjectToStringMap(clusterResponse), "FIELD", "VALUE");
        } catch (IndexOutOfBoundsException ex) {
            throw exceptionTransformer.transformToRuntimeException("There was no cluster for this account.");
        } catch (Exception ex) {
            throw exceptionTransformer.transformToRuntimeException(ex);
        }
    }

    @CliCommand(value = "cluster node --ADD", help = "Add new nodes to the cluster")
    public String addNodeToCluster(
            @CliOption(key = "hostgroup", mandatory = true, help = "Name of the hostgroup") HostGroup hostGroup,
            @CliOption(key = "adjustment", mandatory = true, help = "Count of the nodes which will be added to the cluster") Integer adjustment) {
        try {
            if (adjustment < 1) {
                return "The adjustment value in case of node addition should be at least 1.";
            }
            UpdateClusterJson updateClusterJson = new UpdateClusterJson();
            HostGroupAdjustmentJson hostGroupAdjustmentJson = new HostGroupAdjustmentJson();
            hostGroupAdjustmentJson.setScalingAdjustment(adjustment);
            hostGroupAdjustmentJson.setWithStackUpdate(false);
            hostGroupAdjustmentJson.setHostGroup(hostGroup.getName());
            updateClusterJson.setHostGroupAdjustment(hostGroupAdjustmentJson);
            String stackId = context.isMarathonMode() ? marathonContext.getSelectedMarathonStackId().toString() : context.getStackId();
            cloudbreakClient.clusterEndpoint().put(Long.valueOf(stackId), updateClusterJson);
            return context.getStackId();
        } catch (Exception ex) {
            throw exceptionTransformer.transformToRuntimeException(ex);
        }
    }

    @CliCommand(value = "cluster node --REMOVE", help = "Remove nodes from the cluster")
    public String removeNodeToCluster(
            @CliOption(key = "hostgroup", mandatory = true, help = "Name of the hostgroup") HostGroup hostGroup,
            @CliOption(key = "adjustment", mandatory = true, help = "The number of the nodes to be removed from the cluster.") Integer adjustment,
            @CliOption(key = "withStackDownScale", mandatory = false, help = "Do the downscale with the stack together") Boolean withStackDownScale) {
        try {
            if (adjustment > -1) {
                return "The adjustment value in case of node removal should be negative.";
            }
            UpdateClusterJson updateClusterJson = new UpdateClusterJson();
            HostGroupAdjustmentJson hostGroupAdjustmentJson = new HostGroupAdjustmentJson();
            hostGroupAdjustmentJson.setScalingAdjustment(adjustment);
            hostGroupAdjustmentJson.setWithStackUpdate(withStackDownScale == null ? false : withStackDownScale);
            hostGroupAdjustmentJson.setHostGroup(hostGroup.getName());
            updateClusterJson.setHostGroupAdjustment(hostGroupAdjustmentJson);
            String stackId = context.isMarathonMode() ? marathonContext.getSelectedMarathonStackId().toString() : context.getStackId();
            cloudbreakClient.clusterEndpoint().put(Long.valueOf(stackId), updateClusterJson);
            return context.getStackId();
        } catch (Exception ex) {
            throw exceptionTransformer.transformToRuntimeException(ex);
        }
    }

    @CliCommand(value = "cluster create", help = "Create a new cluster based on a blueprint and optionally a recipe")
    public String createCluster(
            @CliOption(key = "userName", mandatory = false, unspecifiedDefaultValue = "admin", help = "Username of the Ambari server") String userName,
            @CliOption(key = "password", mandatory = false, unspecifiedDefaultValue = "admin", help = "Password of the Ambari server") String password,
            @CliOption(key = "description", mandatory = false, help = "Description of the blueprint") String description,
            @CliOption(key = "stack", mandatory = false, help = "Stack definition name, like HDP") String stack,
            @CliOption(key = "version", mandatory = false, help = "Stack definition version") String version,
            @CliOption(key = "os", mandatory = false, help = "Stack OS to select package manager, default is RedHat") String os,
            @CliOption(key = "stackRepoId", mandatory = false, help = "Stack repository id") String stackRepoId,
            @CliOption(key = "stackBaseURL", mandatory = false, help = "Stack url") String stackBaseURL,
            @CliOption(key = "utilsRepoId", mandatory = false, help = "Stack utils repoId") String utilsRepoId,
            @CliOption(key = "utilsBaseURL", mandatory = false, help = "Stack utils URL") String utilsBaseURL,
            @CliOption(key = "verify", mandatory = false, help = "Whether to verify the URLs or not") Boolean verify,
            @CliOption(key = "enableSecurity", mandatory = false, specifiedDefaultValue = "true", unspecifiedDefaultValue = "false",
                    help = "Kerberos security status") Boolean enableSecurity,
            @CliOption(key = "kerberosMasterKey", mandatory = false, specifiedDefaultValue = "key", help = "Kerberos mater key") String kerberosMasterKey,
            @CliOption(key = "kerberosAdmin", mandatory = false, specifiedDefaultValue = "admin", help = "Kerberos admin name") String kerberosAdmin,
            @CliOption(key = "kerberosPassword", mandatory = false, specifiedDefaultValue = "admin", help = "Kerberos admin password") String kerberosPassword,
            @CliOption(key = "ldapRequired", mandatory = false, unspecifiedDefaultValue = "false", specifiedDefaultValue = "true",
                    help = "Start and configure LDAP authentication support for Ambari hosts") Boolean ldapRequired,
            @CliOption(key = "configStrategy", mandatory = false, help = "Config recommendation strategy") ConfigStrategy strategy,
            @CliOption(key = "wait", mandatory = false, help = "Wait for stack creation", specifiedDefaultValue = "false") Boolean wait) {
        try {
            Set<HostGroupJson> hostGroupList = new HashSet<>();
            Set<Map.Entry<String, HostgroupEntry>> entries = context.isMarathonMode()
                    ? marathonContext.getHostgroups().entrySet() : context.getHostGroups().entrySet();
            for (Map.Entry<String, HostgroupEntry> entry : entries) {
                HostGroupJson hostGroupJson = new HostGroupJson();
                hostGroupJson.setRecipeIds(entry.getValue().getRecipeIdSet());
                hostGroupJson.setName(entry.getKey());

                ConstraintJson constraintJson = new ConstraintJson();

                constraintJson.setHostCount(entry.getValue().getNodeCount());
                if (context.isMarathonMode()) {
                    constraintJson.setConstraintTemplateName(entry.getValue().getConstraintName());
                } else {
                    constraintJson.setInstanceGroupName(entry.getKey());
                }

                hostGroupJson.setConstraint(constraintJson);
                hostGroupJson.setRecipeIds(entry.getValue().getRecipeIdSet());
                hostGroupList.add(hostGroupJson);
            }

            wait = wait == null ? false : wait;
            ClusterRequest clusterRequest = new ClusterRequest();
            clusterRequest.setName(context.isMarathonMode() ? marathonContext.getSelectedMarathonStackName() : context.getStackName());
            clusterRequest.setDescription(description);
            clusterRequest.setUserName(userName);
            clusterRequest.setPassword(password);
            clusterRequest.setBlueprintId(Long.valueOf(context.getBlueprintId()));
            clusterRequest.setEmailNeeded(false);
            clusterRequest.setEnableSecurity(enableSecurity);
            clusterRequest.setHostGroups(hostGroupList);

            if (strategy != null) {
                clusterRequest.setConfigStrategy(strategy);
            }

            if (!context.isMarathonMode()) {
                FileSystemRequest fileSystemRequest = new FileSystemRequest();
                fileSystemRequest.setName(context.getStackName());
                fileSystemRequest.setDefaultFs(context.getDefaultFileSystem() == null ? true : context.getDefaultFileSystem());
                fileSystemRequest.setType(context.getFileSystemType());

                if (context.getDefaultFileSystem() == null && context.getFileSystemType() == null) {
                    fileSystemRequest = null;
                }
                clusterRequest.setFileSystem(fileSystemRequest);
            }
            clusterRequest.setKerberosAdmin(kerberosAdmin);
            clusterRequest.setKerberosMasterKey(kerberosMasterKey);
            clusterRequest.setKerberosPassword(kerberosPassword);
            clusterRequest.setLdapRequired(ldapRequired);
            if (context.getSssdConfigId() != null) {
                clusterRequest.setSssdConfigId(Long.valueOf(context.getSssdConfigId()));
            }
            clusterRequest.setValidateBlueprint(false);

            AmbariStackDetailsJson ambariStackDetailsJson = new AmbariStackDetailsJson();
            ambariStackDetailsJson.setOs(os);
            ambariStackDetailsJson.setStack(stack);
            ambariStackDetailsJson.setStackBaseURL(stackBaseURL);
            ambariStackDetailsJson.setStackRepoId(stackRepoId);
            ambariStackDetailsJson.setUtilsBaseURL(utilsBaseURL);
            ambariStackDetailsJson.setUtilsRepoId(utilsRepoId);
            ambariStackDetailsJson.setVerify(verify);
            ambariStackDetailsJson.setVersion(version);

            if (os == null && stack == null && stackBaseURL == null && stackRepoId == null && utilsBaseURL == null
                    && utilsRepoId == null && verify == null && version == null) {
                ambariStackDetailsJson = null;
            }
            clusterRequest.setAmbariStackDetails(ambariStackDetailsJson);

            String stackId = context.isMarathonMode() ? marathonContext.getSelectedMarathonStackId().toString() : context.getStackId();
            cloudbreakClient.clusterEndpoint().post(Long.valueOf(stackId), clusterRequest);
            context.setHint(Hints.NONE);
            context.resetFileSystemConfiguration();
            if (wait) {
                CloudbreakShellUtil.WaitResult waitResult =
                        cloudbreakShellUtil.waitAndCheckClusterStatus(Long.valueOf(stackId), Status.AVAILABLE.name());
                if (CloudbreakShellUtil.WaitResult.FAILED.equals(waitResult)) {
                    throw exceptionTransformer.transformToRuntimeException("Cluster creation failed on stack with id: " + stackId);
                } else {
                    return "Cluster creation finished";
                }
            }
            return "Cluster creation started";
        } catch (Exception ex) {
            throw exceptionTransformer.transformToRuntimeException(ex);
        }
    }

    @CliCommand(value = "cluster stop", help = "Stop your cluster")
    public String stopCluster() {
        try {
            UpdateClusterJson updateClusterJson = new UpdateClusterJson();
            updateClusterJson.setStatus(StatusRequest.STOPPED);
            String stackId = context.isMarathonMode() ? marathonContext.getSelectedMarathonStackId().toString() : context.getStackId();
            cloudbreakClient.clusterEndpoint().put(Long.valueOf(stackId), updateClusterJson);
            return "Cluster is stopping";
        } catch (Exception ex) {
            throw exceptionTransformer.transformToRuntimeException(ex);
        }
    }

    @CliCommand(value = "cluster start", help = "Start your cluster")
    public String startCluster() {
        try {
            UpdateStackJson updateStackJson = new UpdateStackJson();
            updateStackJson.setStatus(StatusRequest.STARTED);
            String stackId = context.isMarathonMode() ? marathonContext.getSelectedMarathonStackId().toString() : context.getStackId();
            cloudbreakClient.stackEndpoint().put(Long.valueOf(stackId), updateStackJson);
            return "Cluster is starting";
        } catch (Exception ex) {
            throw exceptionTransformer.transformToRuntimeException(ex);
        }
    }
}

File: shell/src/main/java/com/sequenceiq/cloudbreak/shell/commands/common/ClusterCommands.kt
package com.sequenceiq.cloudbreak.shell.commands.common

import java.util.HashSet

import org.apache.http.MethodNotSupportedException
import org.springframework.shell.core.annotation.CliAvailabilityIndicator
import org.springframework.shell.core.annotation.CliCommand
import org.springframework.shell.core.annotation.CliOption

import com.sequenceiq.cloudbreak.api.model.AmbariStackDetailsJson
import com.sequenceiq.cloudbreak.api.model.ClusterRequest
import com.sequenceiq.cloudbreak.api.model.ClusterResponse
import com.sequenceiq.cloudbreak.api.model.ConfigStrategy
import com.sequenceiq.cloudbreak.api.model.ConstraintJson
import com.sequenceiq.cloudbreak.api.model.FileSystemRequest
import com.sequenceiq.cloudbreak.api.model.HostGroupAdjustmentJson
import com.sequenceiq.cloudbreak.api.model.HostGroupJson
import com.sequenceiq.cloudbreak.api.model.RDSConfigJson
import com.sequenceiq.cloudbreak.api.model.RDSDatabase
import com.sequenceiq.cloudbreak.api.model.Status
import com.sequenceiq.cloudbreak.api.model.StatusRequest
import com.sequenceiq.cloudbreak.api.model.UpdateClusterJson
import com.sequenceiq.cloudbreak.shell.commands.BaseCommands
import com.sequenceiq.cloudbreak.shell.completion.HostGroup
import com.sequenceiq.cloudbreak.shell.model.Hints
import com.sequenceiq.cloudbreak.shell.model.HostgroupEntry
import com.sequenceiq.cloudbreak.shell.model.MarathonHostgroupEntry
import com.sequenceiq.cloudbreak.shell.model.NodeCountEntry
import com.sequenceiq.cloudbreak.shell.model.ShellContext
import com.sequenceiq.cloudbreak.shell.util.CloudbreakShellUtil

class ClusterCommands(private val shellContext: ShellContext, private val cloudbreakShellUtil: CloudbreakShellUtil) : BaseCommands {

    @CliAvailabilityIndicator(value = "cluster create")
    fun createAvailable(): Boolean {
        return shellContext.isBlueprintAvailable && (shellContext.isStackAvailable && shellContext.activeHostGroups.size == shellContext.hostGroups.keys.size || shellContext.isMarathonMode && shellContext.isSelectedMarathonStackAvailable
                && shellContext.activeHostGroups.size == shellContext.marathonHostGroups.size)
    }

    @CliCommand(value = "cluster create", help = "Create a new cluster based on a blueprint and optionally a recipe")
    fun createCluster(
            @CliOption(key = "userName", mandatory = false, unspecifiedDefaultValue = "admin", help = "Username of the Ambari server") userName: String,
            @CliOption(key = "password", mandatory = false, unspecifiedDefaultValue = "admin", help = "Password of the Ambari server") password: String,
            @CliOption(key = "description", mandatory = false, help = "Description of the blueprint") description: String,
            @CliOption(key = "stack", mandatory = false, help = "Stack definition name, like HDP") stack: String?,
            @CliOption(key = "version", mandatory = false, help = "Stack definition version") version: String?,
            @CliOption(key = "os", mandatory = false, help = "Stack OS to select package manager, default is RedHat") os: String?,
            @CliOption(key = "stackRepoId", mandatory = false, help = "Stack repository id") stackRepoId: String?,
            @CliOption(key = "stackBaseURL", mandatory = false, help = "Stack url") stackBaseURL: String?,
            @CliOption(key = "utilsRepoId", mandatory = false, help = "Stack utils repoId") utilsRepoId: String?,
            @CliOption(key = "utilsBaseURL", mandatory = false, help = "Stack utils URL") utilsBaseURL: String?,
            @CliOption(key = "verify", mandatory = false, help = "Whether to verify the URLs or not") verify: Boolean?,
            @CliOption(key = "connectionURL", mandatory = false, help = "JDBC connection URL (jdbc:<db-type>://<address>:<port>/<db>)") connectionURL: String?,
            @CliOption(key = "databaseType", mandatory = false, help = "Type of the external database (MYSQL, POSTGRES)") databaseType: RDSDatabase?,
            @CliOption(key = "connectionUserName", mandatory = false, help = "Username to use for the jdbc connection") connectionUserName: String?,
            @CliOption(key = "connectionPassword", mandatory = false, help = "Password to use for the jdbc connection") connectionPassword: String?,
            @CliOption(key = "enableSecurity", mandatory = false, specifiedDefaultValue = "true", unspecifiedDefaultValue = "false", help = "Kerberos security status")
            enableSecurity: Boolean?,
            @CliOption(key = "kerberosMasterKey", mandatory = false, specifiedDefaultValue = "key", help = "Kerberos mater key") kerberosMasterKey: String,
            @CliOption(key = "kerberosAdmin", mandatory = false, specifiedDefaultValue = "admin", help = "Kerberos admin name") kerberosAdmin: String,
            @CliOption(key = "kerberosPassword", mandatory = false, specifiedDefaultValue = "admin", help = "Kerberos admin password") kerberosPassword: String,
            @CliOption(key = "ldapRequired", mandatory = false, unspecifiedDefaultValue = "false", specifiedDefaultValue = "true", help = "Start and configure LDAP authentication support for Ambari hosts")
            ldapRequired: Boolean?,
            @CliOption(key = "configStrategy", mandatory = false, help = "Config recommendation strategy") strategy: ConfigStrategy?,
            @CliOption(key = "enableShipyard", mandatory = false, help = "Run shipyard in cluster") enableShipyard: Boolean?,
            @CliOption(key = "wait", mandatory = false, help = "Wait for stack creation", specifiedDefaultValue = "false") wait: Boolean?): String {
        var wait = wait
        try {
            val hostGroupList = HashSet<HostGroupJson>()
            for (entry in if (shellContext.isMarathonMode)
                shellContext.marathonHostGroups.entries
            else
                shellContext.hostGroups.entries) {
                val hostGroupJson = HostGroupJson()
                hostGroupJson.name = entry.key

                val constraintJson = ConstraintJson()

                constraintJson.hostCount = entry.value.nodeCount
                if (shellContext.isMarathonMode) {
                    constraintJson.constraintTemplateName = (entry.value as MarathonHostgroupEntry).constraintName
                } else {
                    hostGroupJson.recipeIds = (entry.value as HostgroupEntry).recipeIdSet
                    constraintJson.instanceGroupName = entry.key
                }

                hostGroupJson.constraint = constraintJson
                hostGroupList.add(hostGroupJson)
            }

            wait = if (wait == null) false else wait
            val clusterRequest = ClusterRequest()
            clusterRequest.enableShipyard = enableShipyard ?: false
            clusterRequest.name = if (shellContext.isMarathonMode) shellContext.selectedMarathonStackName else shellContext.stackName
            clusterRequest.description = description
            clusterRequest.userName = userName
            clusterRequest.password = password
            clusterRequest.blueprintId = java.lang.Long.valueOf(shellContext.blueprintId)
            clusterRequest.emailNeeded = false
            clusterRequest.enableSecurity = enableSecurity
            clusterRequest.hostGroups = hostGroupList

            if (strategy != null) {
                clusterRequest.configStrategy = strategy
            }

            if (!shellContext.isMarathonMode) {
                var fileSystemRequest: FileSystemRequest? = FileSystemRequest()
                fileSystemRequest!!.name = shellContext.stackName
                fileSystemRequest.isDefaultFs = if (shellContext.defaultFileSystem == null) true else shellContext.defaultFileSystem
                fileSystemRequest.type = shellContext.fileSystemType

                if (shellContext.defaultFileSystem == null && shellContext.fileSystemType == null) {
                    fileSystemRequest = null
                }
                clusterRequest.fileSystem = fileSystemRequest
            }
            clusterRequest.kerberosAdmin = kerberosAdmin
            clusterRequest.kerberosMasterKey = kerberosMasterKey
            clusterRequest.kerberosPassword = kerberosPassword
            clusterRequest.ldapRequired = ldapRequired
            if (shellContext.sssdConfigId != null) {
                clusterRequest.sssdConfigId = java.lang.Long.valueOf(shellContext.sssdConfigId)
            }
            clusterRequest.setValidateBlueprint(false)

            var ambariStackDetailsJson: AmbariStackDetailsJson? = AmbariStackDetailsJson()
            ambariStackDetailsJson!!.os = os
            ambariStackDetailsJson.stack = stack
            ambariStackDetailsJson.stackBaseURL = stackBaseURL
            ambariStackDetailsJson.stackRepoId = stackRepoId
            ambariStackDetailsJson.utilsBaseURL = utilsBaseURL
            ambariStackDetailsJson.utilsRepoId = utilsRepoId
            ambariStackDetailsJson.verify = verify
            ambariStackDetailsJson.version = version

            if (os == null && stack == null && stackBaseURL == null && stackRepoId == null && utilsBaseURL == null
                    && utilsRepoId == null && verify == null && version == null) {
                ambariStackDetailsJson = null
            }
            clusterRequest.ambariStackDetails = ambariStackDetailsJson

            if (connectionURL != null && connectionUserName != null && connectionPassword != null && databaseType != null) {
                val rdsConfigJson = RDSConfigJson()
                rdsConfigJson.connectionURL = connectionURL
                rdsConfigJson.databaseType = databaseType
                rdsConfigJson.connectionUserName = connectionUserName
                rdsConfigJson.connectionPassword = connectionPassword
                clusterRequest.rdsConfigJson = rdsConfigJson
            } else if (connectionURL != null || connectionUserName != null || connectionPassword != null || databaseType != null) {
                return "connectionURL, databaseType, connectionUserName and connectionPassword must be all set."
            }

            val stackId = if (shellContext.isMarathonMode) shellContext.selectedMarathonStackId!!.toString() else shellContext.stackId
            cloudbreakShellUtil.checkResponse("createCluster", shellContext.cloudbreakClient().clusterEndpoint().post(java.lang.Long.valueOf(stackId), clusterRequest))
            shellContext.setHint(Hints.NONE)
            shellContext.resetFileSystemConfiguration()
            if (wait) {
                val waitResult = cloudbreakShellUtil.waitAndCheckClusterStatus(java.lang.Long.valueOf(stackId), Status.AVAILABLE.name)
                if (CloudbreakShellUtil.WaitResultStatus.FAILED == waitResult.waitResultStatus) {
                    throw shellContext.exceptionTransformer().transformToRuntimeException(
                            String.format("Cluster creation failed on stack with id: '%s': '%s'", stackId, waitResult.reason))
                } else {
                    return "Cluster creation finished"
                }
            }
            return "Cluster creation started"
        } catch (ex: Exception) {
            throw shellContext.exceptionTransformer().transformToRuntimeException(ex)
        }

    }

    @CliAvailabilityIndicator(value = *arrayOf("cluster stop", "cluster start"))
    fun startStopAvailable(): Boolean {
        return shellContext.isStackAvailable || shellContext.isMarathonMode && shellContext.isSelectedMarathonStackAvailable
    }

    @CliCommand(value = "cluster stop", help = "Stop your cluster")
    fun stop(): String {
        try {
            val updateClusterJson = UpdateClusterJson()
            updateClusterJson.status = StatusRequest.STOPPED
            val stackId = if (shellContext.isMarathonMode) shellContext.selectedMarathonStackId!!.toString() else shellContext.stackId
            cloudbreakShellUtil.checkResponse("stopCluster",
                    shellContext.cloudbreakClient().clusterEndpoint().put(java.lang.Long.valueOf(stackId), updateClusterJson))
            return "Cluster is stopping"
        } catch (ex: Exception) {
            throw shellContext.exceptionTransformer().transformToRuntimeException(ex)
        }

    }

    @CliCommand(value = "cluster start", help = "Start your cluster")
    fun start(): String {
        try {
            val updateClusterJson = UpdateClusterJson()
            updateClusterJson.status = StatusRequest.STARTED
            val stackId = if (shellContext.isMarathonMode) shellContext.selectedMarathonStackId!!.toString() else shellContext.stackId
            cloudbreakShellUtil.checkResponse("startCluster",
                    shellContext.cloudbreakClient().clusterEndpoint().put(java.lang.Long.valueOf(stackId), updateClusterJson))
            return "Cluster is starting"
        } catch (ex: Exception) {
            throw shellContext.exceptionTransformer().transformToRuntimeException(ex)
        }

    }

    override fun deleteAvailable(): Boolean {
        return false
    }

    @Throws(Exception::class)
    override fun delete(id: Long?, name: String?): String {
        throw MethodNotSupportedException("Cluster delete command not available")
    }

    @CliCommand(value = "cluster delete", help = "Delete the cluster by stack id")
    fun delete(): String {
        try {
            val stackId = if (shellContext.isMarathonMode) shellContext.selectedMarathonStackId!!.toString() else shellContext.stackId
            shellContext.cloudbreakClient().clusterEndpoint().delete(java.lang.Long.valueOf(stackId))
            return "Cluster deletion started with stack id: " + stackId
        } catch (ex: Exception) {
            throw shellContext.exceptionTransformer().transformToRuntimeException(ex)
        }

    }

    @Throws(Exception::class)
    override fun deleteById(id: Long?): String {
        return delete(id, null)
    }

    @Throws(Exception::class)
    override fun deleteByName(name: String): String {
        return delete(null, name)
    }

    override fun selectAvailable(): Boolean {
        return false
    }

    @Throws(Exception::class)
    override fun select(id: Long?, name: String?): String {
        throw MethodNotSupportedException("Cluster select command not available")
    }

    @Throws(Exception::class)
    override fun selectById(id: Long?): String {
        return select(id, null)
    }

    @Throws(Exception::class)
    override fun selectByName(name: String): String {
        return select(null, name)
    }

    override fun listAvailable(): Boolean {
        return false
    }

    @Throws(Exception::class)
    override fun list(): String {
        throw MethodNotSupportedException("Cluster list command not available")
    }

    @CliAvailabilityIndicator(value = *arrayOf("cluster show --id", "cluster show --name"))
    override fun showAvailable(): Boolean {
        return shellContext.isStackAvailable || shellContext.isMarathonMode && shellContext.isSelectedMarathonStackAvailable
    }

    @CliCommand(value = "cluster show", help = "Shows the cluster by stack id")
    fun show(): String {
        try {
            val stackId = if (shellContext.isMarathonMode) shellContext.selectedMarathonStackId!!.toString() else shellContext.stackId
            val clusterResponse = shellContext.cloudbreakClient().clusterEndpoint()[java.lang.Long.valueOf(stackId)]
            return shellContext.outputTransformer().render(shellContext.responseTransformer().transformObjectToStringMap(clusterResponse), "FIELD", "VALUE")
        } catch (ex: IndexOutOfBoundsException) {
            throw shellContext.exceptionTransformer().transformToRuntimeException("There was no cluster for this account.")
        } catch (ex: Exception) {
            throw shellContext.exceptionTransformer().transformToRuntimeException(ex)
        }

    }

    @Throws(Exception::class)
    override fun show(id: Long?, name: String?): String {
        throw MethodNotSupportedException("Cluster show command not available")
    }

    @Throws(Exception::class)
    override fun showById(id: Long?): String {
        return show(id, null)
    }

    @Throws(Exception::class)
    override fun showByName(name: String): String {
        return show(null, name)
    }

    @CliAvailabilityIndicator(value = *arrayOf("cluster node"))
    fun nodeAvailable(): Boolean {
        return shellContext.isStackAvailable || shellContext.isMarathonMode && shellContext.isSelectedMarathonStackAvailable
    }

    @CliAvailabilityIndicator(value = *arrayOf("cluster fileSystem"))
    fun fileSystemAvailable(): Boolean {
        return shellContext.isStackAvailable && !shellContext.isMarathonMode
    }

    @CliCommand(value = "cluster node --ADD", help = "Add new nodes to the cluster")
    fun addNode(
            @CliOption(key = "hostgroup", mandatory = true, help = "Name of the hostgroup") hostGroup: HostGroup,
            @CliOption(key = "adjustment", mandatory = true, help = "Count of the nodes which will be added to the cluster") adjustment: Int?): String {
        try {
            if (adjustment < 1) {
                return "The adjustment value in case of node addition should be at least 1."
            }
            val updateClusterJson = UpdateClusterJson()
            val hostGroupAdjustmentJson = HostGroupAdjustmentJson()
            hostGroupAdjustmentJson.scalingAdjustment = adjustment
            hostGroupAdjustmentJson.withStackUpdate = false
            hostGroupAdjustmentJson.hostGroup = hostGroup.name
            updateClusterJson.hostGroupAdjustment = hostGroupAdjustmentJson
            val stackId = if (shellContext.isMarathonMode) shellContext.selectedMarathonStackId!!.toString() else shellContext.stackId
            cloudbreakShellUtil.checkResponse("upscaleCluster",
                    shellContext.cloudbreakClient().clusterEndpoint().put(java.lang.Long.valueOf(stackId), updateClusterJson))
            return shellContext.stackId
        } catch (ex: Exception) {
            throw shellContext.exceptionTransformer().transformToRuntimeException(ex)
        }

    }

    @CliCommand(value = "cluster node --REMOVE", help = "Remove nodes from the cluster")
    fun removeNode(
            @CliOption(key = "hostgroup", mandatory = true, help = "Name of the hostgroup") hostGroup: HostGroup,
            @CliOption(key = "adjustment", mandatory = true, help = "The number of the nodes to be removed from the cluster.") adjustment: Int?,
            @CliOption(key = "withStackDownScale", mandatory = false, help = "Do the downscale with the stack together") withStackDownScale: Boolean?): String {
        try {
            if (adjustment > -1) {
                return "The adjustment value in case of node removal should be negative."
            }
            val updateClusterJson = UpdateClusterJson()
            val hostGroupAdjustmentJson = HostGroupAdjustmentJson()
            hostGroupAdjustmentJson.scalingAdjustment = adjustment
            hostGroupAdjustmentJson.withStackUpdate = withStackDownScale ?: false
            hostGroupAdjustmentJson.hostGroup = hostGroup.name
            updateClusterJson.hostGroupAdjustment = hostGroupAdjustmentJson
            val stackId = if (shellContext.isMarathonMode) shellContext.selectedMarathonStackId!!.toString() else shellContext.stackId
            cloudbreakShellUtil.checkResponse("downscaleCluster",
                    shellContext.cloudbreakClient().clusterEndpoint().put(java.lang.Long.valueOf(stackId), updateClusterJson))
            return shellContext.stackId
        } catch (ex: Exception) {
            throw shellContext.exceptionTransformer().transformToRuntimeException(ex)
        }

    }

    @CliAvailabilityIndicator(value = "cluster sync")
    fun syncAvailable(): Boolean {
        return shellContext.isStackAvailable && !shellContext.isMarathonMode
    }

    @CliCommand(value = "cluster sync", help = "Sync the cluster")
    fun sync(): String {
        try {
            val updateClusterJson = UpdateClusterJson()
            updateClusterJson.status = StatusRequest.SYNC
            shellContext.cloudbreakClient().clusterEndpoint().put(java.lang.Long.valueOf(shellContext.stackId), updateClusterJson)
            return "Cluster is syncing"
        } catch (ex: Exception) {
            throw shellContext.exceptionTransformer().transformToRuntimeException(ex)
        }

    }

    override fun shellContext(): ShellContext {
        return shellContext
    }

}
